libcxx/include/__functional/move_only_function_impl.h

```diff
@@ -0,0 +1,400 @@
+
+  struct _TriviallyDestructibleObject {
+    _CallFn* __call_;
+    alignas(void*) std::byte __data_[__target_size - 1 * sizeof(void*) - 1];
+    _Status __status_;
+  } __trivially_destructible_;
+  static_assert(sizeof(_TriviallyDestructibleObject) == __target_size);
+  static_assert(offsetof(_TriviallyDestructibleObject, __status_) == __target_size - 1);
+
+private:
+  // try to squeeze in another char to see if we can use the last byte for our own purposes
+  template <class _Func>
+  struct _Compact {
+    _LIBCPP_NO_UNIQUE_ADDRESS _Func __func;
+    char __c;
+  };
+
+  template <class _Func, size_t _BufferSize, size_t BufferAlignment>
+  static constexpr bool __fits_in_buffer =
+      sizeof(_Compact<_Func>) <= _BufferSize + 1 && alignof(_Func) <= BufferAlignment;
+
+  template <class _Func>
+  static constexpr bool __fits_in_trivially_relocatable_buffer_impl =
+      __libcpp_is_trivially_relocatable_v<_Func> &&
+      __fits_in_buffer<_Func,
+                       sizeof(_TriviallyRelocatableObject::__data_),
+                       alignof(_TriviallyRelocatableObject::__data_)>;
+
+  template <class _Func>
+  static constexpr bool __fits_in_trivially_destructible_buffer_impl =
+      __libcpp_is_trivially_relocatable_v<_Func> && is_trivially_destructible_v<_Func> &&
+      __fits_in_buffer<_Func,
+                       sizeof(_TriviallyDestructibleObject::__data_),
+                       alignof(_TriviallyDestructibleObject::__data_)>;
+
+public:
+  template <class _Func>
+  static constexpr bool __fits_in_trivially_relocatable_buffer =
+      __fits_in_trivially_relocatable_buffer_impl<remove_cvref_t<_Func>>;
+
+  template <class _Func>
+  static constexpr bool __fits_in_trivially_destructible_buffer =
+      __fits_in_trivially_destructible_buffer_impl<remove_cvref_t<_Func>>;
+};
+
+static_assert(alignof(__move_only_function_storage) == alignof(void*));
+static_assert(sizeof(__move_only_function_storage) == __move_only_function_storage::__target_size);
+
+template <class...>
+class move_only_function;
+
+template <class...>
+struct __is_move_only_function : false_type {};
+
+template <class _ReturnT, class... _ArgTypes>
+struct __is_move_only_function<move_only_function<_ReturnT(_ArgTypes...)>> : true_type {};
+
+_LIBCPP_END_NAMESPACE_STD
+
+// This header is only partially guarded on purpose. This header is an implementation detail of move_only_function.h
+// and generates multiple versions of std::move_only_function
+#endif // _LIBCPP___FUNCTIONAL_MOVE_ONLY_FUNCTION_IMPL_H
+
+#ifndef _LIBCPP_MOVE_ONLY_FUNCTION_CV
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_CV
+#endif
+
+#ifndef _LIBCPP_MOVE_ONLY_FUNCTION_REF
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_REF
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_INV_QUALS _LIBCPP_MOVE_ONLY_FUNCTION_CV&
+#else
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_INV_QUALS _LIBCPP_MOVE_ONLY_FUNCTION_CV _LIBCPP_MOVE_ONLY_FUNCTION_REF
+#endif
```

Comment:
Nit: I think I've intuited that `INV` stands for `INVOKE`, like, the qualifiers that we apply to the callable type right before `std::invoke`'ing it? It took me a while, though. Suggest adding the missing `OKE`, at least. Perhaps even better: `_LIBCPP_MOVE_ONLY_FUNCTION_CVREF_FOR_INVOKE`.

Bikeshed: I'd also remove the underscore from `_LIBCPP_MOVE_ONLY_FUNCTION_CV_REF`, for consistency with e.g. `std::remove_cvref_t`. `cvref` is one word as far as STL stuff is concerned.

The distinction between `_LIBCPP_MOVE_ONLY_FUNCTION_CVREF_FOR_INVOKE` and `_LIBCPP_MOVE_ONLY_FUNCTION_CVREF` seems to be that when `_LIBCPP_MOVE_ONLY_FUNCTION_CVREF` is ref-unqualified, then `_LIBCPP_MOVE_ONLY_FUNCTION_CVREF_FOR_INVOKE` is lvalue-ref-qualified; otherwise they are identical. Right? I wish there were a nice clean metaprogramming-based way to compute `_LIBCPP_MOVE_ONLY_FUNCTION_CVREF_FOR_INVOKE`, instead of needing to do it with the preprocessor. Can you think of one?